backend/src/wamp.ts

```diff
@@ -0,0 +1,557 @@
+    return await receipts.getIncludedReceiptsList(blockHash);
+  },
+  "executed-receipts-list-by-block-hash": async ([blockHash]) => {
+    return await receipts.getExecutedReceiptsList(blockHash);
+  },
+
+  // transactions
+  "is-transaction-indexed": async ([transactionHash]) => {
+    return await transactions.getIsTransactionIndexed(transactionHash);
+  },
+
+  // contracts
+  "contract-info-by-account-id": async ([accountId]) => {
+    return await contracts.getContractInfo(accountId);
+  },
+
+  // chunks
+  "gas-used-in-chunks": async ([blockHash]) => {
+    return await chunks.getGasUsedInChunks(blockHash);
+  },
+};
+
+// set up wamp
+function setupWamp(): () => Promise<autobahn.Session> {
+  const wamp = new autobahn.Connection({
+    realm: "near-explorer",
+    transports: [
+      {
+        url: wampNearExplorerUrl,
+        type: "websocket",
+      },
+    ],
+    authmethods: ["ticket"],
+    authid: "near-explorer-backend",
+    onchallenge: (_session, method) => {
+      if (method === "ticket") {
+        return wampNearExplorerBackendSecret;
+      }
+      throw "WAMP authentication error: unsupported challenge method";
+    },
+    retry_if_unreachable: true,
+    max_retries: Number.MAX_SAFE_INTEGER,
+    max_retry_delay: 10,
+  });
+
+  let currentSessionPromise: Promise<autobahn.Session>;
+  const openEventEmitter = new EventEmitter();
+
+  wamp.onopen = async (session) => {
+    openEventEmitter.emit("opened", session);
+    currentSessionPromise = Promise.resolve(session);
+    console.log("WAMP connection is established. Waiting for commands...");
+
+    for (const [name, handler] of Object.entries(wampHandlers)) {
+      const uri = `com.nearprotocol.${wampNearNetworkName}.explorer.${name}`;
+      try {
+        await session.register(
+          uri,
+          (handler as unknown) as autobahn.RegisterEndpoint
+        );
+      } catch (error) {
+        console.error(`Failed to register "${uri}" handler due to:`, error);
+        wamp.close();
+        setTimeout(() => {
+          wamp.open();
+        }, 1000);
+        return;
+      }
+    }
+  };
+
+  wamp.onclose = (reason) => {
+    currentSessionPromise = new Promise((resolve) => {
+      openEventEmitter.addListener("opened", resolve);
+    });
+    console.log(
+      "WAMP connection has been closed (check WAMP router availability and credentials):",
+      reason
+    );
+    return false;
+  };
+
+  wamp.open();
+  return () => currentSessionPromise;
+}
+
+const wampPublish = async <T extends SubscriptionTopicType>(
+  topic: T,
+  namedArgs: SubscriptionTopicTypes[T],
+  getSession: () => Promise<autobahn.Session>
+): Promise<void> => {
+  const uri = `com.nearprotocol.${wampNearNetworkName}.explorer.${topic}`;
+  const session = await getSession();
+  if (!session.isOpen) {
+    console.log(`No session on stack\n${new Error().stack}`);
+  }
+  session.publish(uri, [], namedArgs);
```

Comment:
We'll get an error "unreachable". I'll put a return statement here.

Reply:
@luixo I think we should do the following:

```
try {
   publish()
} catch {
   if (!session.isOpen) {
      console.log...
   }
   throw
}
```

As otherwise there will be a chance of a race condition

Reply:
How come?
Is there any async code inside `publish` that will try to use session?
I'll add try-catch block around the whole thing just in case, but I believe it's not the case.